### PR TITLE
support i386

### LINKED ISSUE #341: Support 32-bit builds
When I try to run it in my 32-bit build environment, I get the following error:

```
arch/lkl/Makefile:14: *** Unrecognized platform: elf32-i386.  Stop.
```

Not sure if it's something weird about my environment (which is admittedly unusual), but I figured I'd avoid wasting time on it if it's not expected to work on 32-bit systems in the first place.

### PR DESCRIPTION
**Pull request by luckyluke**
Basic support for i386, should fix #341

---
This change is [Reviewable](https://reviewable.io/reviews/lkl/linux/456)

### CODE REVIEW

File: arch/lkl/Makefile [hunk 1]
```diff
@@ -24,7 +24,13 @@ endif
 
 LDFLAGS_vmlinux += -r
 LKL_ENTRY_POINTS := lkl_start_kernel lkl_sys_halt lkl_syscall lkl_trigger_irq \
-	lkl_get_free_irq lkl_put_irq lkl_is_running
+	lkl_get_free_irq lkl_put_irq lkl_is_running lkl_bug lkl_printf
+
+ifeq ($(OUTPUT_FORMAT),elf32-i386)
+LKL_ENTRY_POINTS += \
```
[the remark this answers is not in the document]
Comment (luckyluke) [reply arch/lkl/Makefile]: ok, i'll update the patch in the next days

**Comment by aostrouhhov** [comment]
It should be mentioned that, I suppose, different compiler versions are looking for different symbol names: sometimes it is ```__x86.get_pc_thunk.REG```, sometimes ```__i386.get_pc_thunk.REG``` or ```__i686.get_pc_thunk.REG```. I dodn't know how many of them there are.

**Comment by luckyluke** [comment]
> It should be mentioned that, I suppose, different compiler versions are looking for different symbol names

Thank you, I didn't think about that. Any idea how to detect the compiler? From a quick grep I didn't find examples in linux, maybe something like
`ifneq ($(shell $(CC) --version | grep gcc),)`
to select which compiler-specific symbols to keep?

**Comment by luckyluke** [comment]
I tried with 32-bit clang and no modifications are necessary, everything compiles and test pass as well. Unfortunately I can't test other compilers like intel icc.

**Comment by luckyluke** [comment]
removed the commit with the LKL_ENTRY_POINTS modification

**Comment by thehajime** [comment]
thanks @luckyluke. the update is fine.

another thought is if we can have CI (circleci) for i386 build/tests. it would be great if you added an i386 build instance for circleci if possible ?

current our CI is running here:

https://circleci.com/gh/lkl/workflows/linux/tree/master

**Comment by luckyluke** [comment]
Hi @thehajime, thank you for the feedback.
I'll try to add the circleci conf in the next days.

**Comment by luckyluke** [comment]
Hi @thehajime I added a circleci config on a container I created based on ubuntu i386. This required some more fix for the tests, I guess due to different versions of gcc and btrfs-tools from my debian machine.

**Comment by thehajime** [comment]
> Hi @thehajime I added a circleci config on a container I created based on ubuntu i386. 

Thanks !  LGTM.

> This required some more fix for the tests, I guess due to different versions of gcc and btrfs-tools from my debian machine.

for the btrfs-tools, it is also useful with newer mkfs.btrfs; my fedora uses `btrfs-progs v4.17.1` and this patch fix the issue so, 1589abbeb763b5c3754f9408a0315d17c14858a4 is okay.

**Comment by thehajime** [comment]
@luckyluke for the follow up fix, could you please make a pull request to https://github.com/lkl/lkl-docker/commits/master ? if we could make the image, we may be able to pull the docker image from docker hub (lkldocker/).

**Pull request merged**